File: android/app/src/main/java/com/laneshadow/ui/components/atoms/DeviationPolyline.kt
package com.laneshadow.ui.components.atoms

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.LocalInspectionMode
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * Geographic coordinate point
 *
 * Represents a latitude/longitude coordinate on Earth
 */
data class LatLng(
    val latitude: Double,
    val longitude: Double,
)

/**
 * DeviationPolyline component
 *
 * Displays route deviation information with original route, detour path, and reconnect point.
 * This is a map component that currently renders as a Canvas placeholder for demonstration.
 * When Google Maps SDK is integrated, this will wire into the actual map rendering.
 *
 * Following RN wrapper API from react-native/components/ui/map/deviation-polyline.tsx
 *
 * @param originalRoute List of LatLng points representing the original planned route
 * @param detourPath List of LatLng points representing the detour/alternate path
 * @param reconnectPoint Optional LatLng point where detour reconnects to original route
 * @param strokeWidth Width of the route lines in dp (default 6dp)
 * @param onPress Optional callback when the polyline is pressed
 * @param testID Test ID for UI testing
 * @param modifier Modifier for the component container
 */
@Composable
fun DeviationPolyline(
    originalRoute: List<LatLng>,
    detourPath: List<LatLng>,
    reconnectPoint: LatLng? = null,
    strokeWidth: Dp = 6.dp,
    onPress: (() -> Unit)? = null,
    testID: String? = null,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current
    val isPreview = LocalInspectionMode.current

    // Map deviation colors from theme domain colors
    val originalRouteColor: Color = theme.domain.deviationOriginalRoute.default
    val detourPathColor: Color = theme.domain.deviationDetourPath.default
    val reconnectPointColor: Color = theme.domain.deviationReconnectPoint.default

    // Build content description for accessibility
    val contentDescription = buildDeviationContentDescription(
        originalRoute.size,
        detourPath.size,
        reconnectPoint != null,
    )

    Canvas(
        modifier = modifier
            .fillMaxSize()
            .semantics {
                this.contentDescription = contentDescription
            },
    ) {
        val strokeWidthPx = strokeWidth.value * density

        // Draw original route (gray, solid line)
        if (originalRoute.size >= 2) {
            val originalPath = createPathFromPoints(originalRoute, size.width, size.height)
            drawPath(
                path = originalPath,
                color = originalRouteColor,
                style = Stroke(
                    width = strokeWidthPx,
                    pathEffect = PathEffect.dashPathEffect(floatArrayOf(10f, 10f), 0f),
                ),
            )
        }

        // Draw detour path (orange, solid line)
        if (detourPath.size >= 2) {
            val detourPath = createPathFromPoints(detourPath, size.width, size.height)
            drawPath(
                path = detourPath,
                color = detourPathColor,
                style = Stroke(width = strokeWidthPx),
            )
        }

        // Draw reconnect point (green circle)
        if (reconnectPoint != null) {
            val reconnectOffset = latLngToOffset(
                reconnectPoint,
                originalRoute + detourPath,
                size.width,
                size.height,
            )
            drawCircle(
                color = reconnectPointColor,
                radius = strokeWidthPx * 1.5f,
                center = reconnectOffset,
            )
        }
    }
}

/**
 * Build accessibility content description for the deviation
 */
private fun buildDeviationContentDescription(
    originalPointCount: Int,
    detourPointCount: Int,
    hasReconnectPoint: Boolean,
): String {
    return buildString {
        append("Route deviation. ")
        append("Original route has $originalPointCount points. ")
        append("Detour path has $detourPointCount points. ")
        if (hasReconnectPoint) {
            append("Reconnect point marked.")
        }
    }
}

/**
 * Create a Compose Path from a list of LatLng points
 *
 * This is a simplified implementation that maps geographic coordinates
 * to canvas coordinates for demonstration purposes.
 * In production with Google Maps SDK, this would be handled by the map platform.
 */
private fun createPathFromPoints(
    points: List<LatLng>,
    canvasWidth: Float,
    canvasHeight: Float,
): Path {
    val path = Path()

    if (points.isEmpty()) return path

    // Find bounding box of all points
    val minLat = points.minOfOrNull { it.latitude } ?: 0.0
    val maxLat = points.maxOfOrNull { it.latitude } ?: 0.0
    val minLon = points.minOfOrNull { it.longitude } ?: 0.0
    val maxLon = points.maxOfOrNull { it.longitude } ?: 0.0

    val latRange = maxLat - minLat
    val lonRange = maxLon - minLon

    // Map first point
    val firstOffset = latLngToOffset(
        points[0],
        points,
        canvasWidth,
        canvasHeight,
    )
    path.moveTo(firstOffset.x, firstOffset.y)

    // Map remaining points
    for (i in 1 until points.size) {
        val offset = latLngToOffset(
            points[i],
            points,
            canvasWidth,
            canvasHeight,
        )
        path.lineTo(offset.x, offset.y)
    }

    return path
}

/**
 * Convert a LatLng point to canvas Offset
 *
 * This is a simplified projection that maps geographic coordinates
 * to canvas space for demonstration purposes.
 * Uses linear interpolation within the bounding box of all points.
 */
private fun latLngToOffset(
    point: LatLng,
    allPoints: List<LatLng>,
    canvasWidth: Float,
    canvasHeight: Float,
): Offset {
    if (allPoints.isEmpty()) {
        return Offset(canvasWidth / 2f, canvasHeight / 2f)
    }

    val minLat = allPoints.minOfOrNull { it.latitude } ?: point.latitude
    val maxLat = allPoints.maxOfOrNull { it.latitude } ?: point.latitude
    val minLon = allPoints.minOfOrNull { it.longitude } ?: point.longitude
    val maxLon = allPoints.maxOfOrNull { it.longitude } ?: point.longitude

    val latRange = maxLat - minLat
    val lonRange = maxLon - minLon

    // Add padding to prevent drawing on edges
    val padding = 20f
    val usableWidth = canvasWidth - 2 * padding
    val usableHeight = canvasHeight - 2 * padding

    val x: Float = if (lonRange > 0) {
        (padding + (point.longitude - minLon) / lonRange * usableWidth).toFloat()
    } else {
        canvasWidth / 2f
    }

    val y: Float = if (latRange > 0) {
        (padding + (1 - (point.latitude - minLat) / latRange) * usableHeight).toFloat()
    } else {
        canvasHeight / 2f
    }

    return Offset.Unspecified.copy(x = x, y = y)
}

File: ios/LaneShadow/Views/Atoms/DeviationPolyline.swift
import LaneShadowTheme
import SwiftUI

/// Public coordinate model for geographic locations
public struct CLLocationCoordinate: Equatable, Sendable {
    public var latitude: Double
    public var longitude: Double

    public init(latitude: Double, longitude: Double) {
        self.latitude = latitude
        self.longitude = longitude
    }
}

/// DeviationPolyline component - Visual representation of route deviations
///
/// Displays the original route, detour path, and reconnection point on a map.
/// Uses Canvas as a visual placeholder for future MapKit integration.
///
/// Following the translation matrix specification:
/// - Original route stroke: theme.domain.deviationOriginalRoute (fallback gray)
/// - Detour path stroke: theme.domain.deviationDetourPath (fallback orange)
/// - Reconnect point: theme.domain.deviationReconnectPoint (fallback green)
/// - Stroke width: 6pt default
public struct DeviationPolyline: View {
    // MARK: - Properties

    @Environment(\.theme) private var theme

    private let originalRoute: [CLLocationCoordinate]
    private let detourPath: [CLLocationCoordinate]
    private let reconnectPoint: CLLocationCoordinate?
    private let strokeWidth: CGFloat
    private let onPress: (() -> Void)?
    private let testID: String?

    // MARK: - Initialization

    /// Creates a DeviationPolyline with the given route data
    /// - Parameters:
    ///   - originalRoute: Array of coordinates representing the original route
    ///   - detourPath: Array of coordinates representing the detour path
    ///   - reconnectPoint: Optional coordinate where the detour reconnects to the original route
    ///   - strokeWidth: Width of the stroke (default is 6pt)
    ///   - onPress: Optional closure to call when the polyline is pressed
    ///   - testID: Optional identifier for testing
    public init(
        originalRoute: [CLLocationCoordinate],
        detourPath: [CLLocationCoordinate],
        reconnectPoint: CLLocationCoordinate? = nil,
        strokeWidth: CGFloat = 6,
        onPress: (() -> Void)? = nil,
        testID: String? = nil
    ) {
        self.originalRoute = originalRoute
        self.detourPath = detourPath
        self.reconnectPoint = reconnectPoint
        self.strokeWidth = strokeWidth
        self.onPress = onPress
        self.testID = testID
    }

    // MARK: - Body

    public var body: some View {
        Canvas { context, size in
            // Draw original route
            drawPolyline(
                context: context,
                coordinates: originalRoute,
                color: theme.domain.deviationOriginalRoute.default,
                size: size
            )

            // Draw detour path
            drawPolyline(
                context: context,
                coordinates: detourPath,
                color: theme.domain.deviationDetourPath.default,
                size: size
            )

            // Draw reconnect point if provided
            if let reconnectPoint {
                drawReconnectPoint(
                    context: context,
                    coordinate: reconnectPoint,
                    size: size
                )
            }
        }
        .frame(height: 200) // Placeholder height for Canvas
        .contentShape(Rectangle())
        .onTapGesture {
            onPress?()
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("Route deviation map")
        .accessibilityHint("Shows original route in gray, detour path in orange, and reconnection point in green")
        .accessibilityAddTraits(.isButton)
        .accessibilityIdentifier(testID ?? "deviationPolyline")
    }

    // MARK: - Private Helpers

    private func drawPolyline(
        context: GraphicsContext,
        coordinates: [CLLocationCoordinate],
        color: Color,
        size: CGSize
    ) {
        guard coordinates.count >= 2 else { return }

        var path = Path()

        // Convert coordinates to normalized canvas coordinates
        let points = coordinates.map { coordinate in
            normalizeCoordinate(coordinate, in: size)
        }

        if let firstPoint = points.first {
            path.move(to: firstPoint)
        }

        for point in points.dropFirst() {
            path.addLine(to: point)
        }

        context.stroke(
            path,
            with: .color(color),
            lineWidth: strokeWidth
        )
    }

    private func drawReconnectPoint(
        context: GraphicsContext,
        coordinate: CLLocationCoordinate,
        size: CGSize
    ) {
        let point = normalizeCoordinate(coordinate, in: size)
        let radius: CGFloat = 8

        var circlePath = Path()
        circlePath.addEllipse(in: CGRect(
            x: point.x - radius,
            y: point.y - radius,
            width: radius * 2,
            height: radius * 2
        ))

        context.fill(
            circlePath,
            with: .color(theme.domain.deviationReconnectPoint.default)
        )
    }

    private func normalizeCoordinate(
        _ coordinate: CLLocationCoordinate,
        in size: CGSize
    ) -> CGPoint {
        // Find bounding box of all coordinates
        let allCoordinates = originalRoute + detourPath + ([reconnectPoint].compactMap { $0 })

        let minLat = allCoordinates.map(\.latitude).min() ?? coordinate.latitude
        let maxLat = allCoordinates.map(\.latitude).max() ?? coordinate.latitude
        let minLon = allCoordinates.map(\.longitude).min() ?? coordinate.longitude
        let maxLon = allCoordinates.map(\.longitude).max() ?? coordinate.longitude

        let latRange = maxLat - minLat
        let lonRange = maxLon - minLon

        // Normalize to canvas size with padding
        let padding: CGFloat = 20
        let usableWidth = size.width - (padding * 2)
        let usableHeight = size.height - (padding * 2)

        let normalizedX: CGFloat
        let normalizedY: CGFloat

        if lonRange > 0 {
            normalizedX = padding + ((coordinate.longitude - minLon) / lonRange) * usableWidth
        } else {
            normalizedX = size.width / 2
        }

        if latRange > 0 {
            normalizedY = padding + ((coordinate.latitude - minLat) / latRange) * usableHeight
        } else {
            normalizedY = size.height / 2
        }

        return CGPoint(x: normalizedX, y: normalizedY)
    }
}

// MARK: - Preview

#Preview("Deviation Polyline") {
    VStack(spacing: 20) {
        DeviationPolyline(
            originalRoute: [
                CLLocationCoordinate(latitude: 37.7749, longitude: -122.4194),
                CLLocationCoordinate(latitude: 37.7750, longitude: -122.4180),
                CLLocationCoordinate(latitude: 37.7755, longitude: -122.4170),
                CLLocationCoordinate(latitude: 37.7765, longitude: -122.4160),
                CLLocationCoordinate(latitude: 37.7775, longitude: -122.4150),
            ],
            detourPath: [
                CLLocationCoordinate(latitude: 37.7750, longitude: -122.4180),
                CLLocationCoordinate(latitude: 37.7752, longitude: -122.4175),
                CLLocationCoordinate(latitude: 37.7760, longitude: -122.4170),
                CLLocationCoordinate(latitude: 37.7765, longitude: -122.4160),
            ],
            reconnectPoint: CLLocationCoordinate(latitude: 37.7765, longitude: -122.4160),
            strokeWidth: 6,
            testID: "deviationPreview"
        )
        .padding()

        Text("Deviation Polyline")
            .font(.headline)
    }
    .laneShadowTheme()
}

#Preview("Without Reconnect Point") {
    VStack(spacing: 20) {
        DeviationPolyline(
            originalRoute: [
                CLLocationCoordinate(latitude: 37.7749, longitude: -122.4194),
                CLLocationCoordinate(latitude: 37.7750, longitude: -122.4180),
                CLLocationCoordinate(latitude: 37.7755, longitude: -122.4170),
            ],
            detourPath: [
                CLLocationCoordinate(latitude: 37.7750, longitude: -122.4180),
                CLLocationCoordinate(latitude: 37.7752, longitude: -122.4175),
                CLLocationCoordinate(latitude: 37.7760, longitude: -122.4170),
            ],
            strokeWidth: 6
        )
        .padding()

        Text("Without Reconnect Point")
            .font(.headline)
    }
    .laneShadowTheme()
}

#Preview("Single Point") {
    VStack(spacing: 20) {
        DeviationPolyline(
            originalRoute: [
                CLLocationCoordinate(latitude: 37.7749, longitude: -122.4194),
            ],
            detourPath: [
                CLLocationCoordinate(latitude: 37.7750, longitude: -122.4180),
            ],
            reconnectPoint: CLLocationCoordinate(latitude: 37.7755, longitude: -122.4170),
            strokeWidth: 6
        )
        .padding()

        Text("Single Point (Degenerate)")
            .font(.headline)
    }
    .laneShadowTheme()
}
